lib/rspec/expectations/failure_aggregator.rb

```diff
@@ -0,0 +1,159 @@
+module RSpec
+  module Expectations
+    # @private
+    class FailureAggregator
+      attr_reader :block_label, :metadata
+
+      def aggregate
+        RSpec::Support.with_failure_notifier(self) do
+          begin
+            yield
+          rescue ExpectationNotMetError => e
+            failures << e
+          rescue Exception => e
```

Comment:
ah, that makes sense. Maybe include some of that in a comment? Seems non-obvious.

Don't think any code changes are needed here though - seems like an edge case. If an exception is raise I think it's reasonable to expect we didn't run all the expectations. And the thread case is pretty weird, I can't imagine people hitting that often.